src/NuGet.Core/NuGet.Packaging/TopologicalSortUtility.cs

```diff
@@ -36,21 +33,68 @@ public static class TopologicalSortUtility
             return sorted;
         }
 
-        private static int GetParentCount(List<PackageDependencyInfo> packages, string id)
+        private static void Sort(PackageInfo[] packages, int start)
+        {
+            // Update parent counts
+            for (var i = start; i < packages.Length; i++)
+            {
+                var package = packages[i].Package;
+                // Mutating stuct in-place, need to index rather than use enumerator
+                packages[i].ParentCount = GetParentCount(packages, package.Id, start);
+            }
+
+            Array.Sort(packages, start, packages.Length - start, _comparer);
+        }
+
+        private static int GetParentCount(PackageInfo[] packages, string id, int start)
         {
-            int count = 0;
+            var parentCount = 0;
 
-            foreach (var package in packages)
+            for (var i = start; i < packages.Length; i++)
             {
-                if (package.Dependencies != null
-                    && package.Dependencies.Any(dependency =>
-                        string.Equals(id, dependency.Id, StringComparison.OrdinalIgnoreCase)))
+                var deps = packages[i].Package.Dependencies;
+                var dependencies = deps as PackageDependency[] ?? deps.ToArray();
```

Comment:
It allocates [behind IEnumerable](https://gist.github.com/benaadams/66f736fd0a0fce5699863e4c9e69e851) and is and order of magnitude slower 
```
                  Method |      Mean |     Error |    StdDev | Allocated |
------------------------ |----------:|----------:|----------:|----------:|
            ForEachArray |  2.219 us | 0.0079 us | 0.0073 us |       0 B |
       ForEachArrayIList | 25.415 us | 0.0829 us | 0.0775 us |      32 B |
 ForEachArrayIEnumerable | 25.453 us | 0.0543 us | 0.0481 us |      32 B |
               LinqArray | 31.325 us | 0.1859 us | 0.1739 us |      32 B |
```
A regular `for` loop with `IList<T>` and a cached length (so it doesn't evaluate the property each loop) is ok (rather than `foreach`) if you want to do an interface e.g.

```csharp
public int ForArrayIList()
{
    var counter = 0;
    var list = _ilist;
    
    var count = list.Count;
    for (var i = 0; i < count; i++)
    {
        counter += i;
    }
    return counter;
}
```